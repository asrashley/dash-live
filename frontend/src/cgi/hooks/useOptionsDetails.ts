import { type ReadonlySignal, useSignal } from "@preact/signals";

import { fieldGroups } from "@dashlive/options";

import { InputOptionName } from "../types/InputOptionName";

function createOptionNames(): InputOptionName[] {
    const names: InputOptionName[] = [];
    fieldGroups.forEach(grp => {
        grp.fields.forEach(({ name, fullName, shortName }) => {
            names.push({
                cgiName: name,
                fullName,
                shortName,
            });
        });
    });
    return names;
}

export interface UseOptionsDetailsHook {
    allOptions: ReadonlySignal<InputOptionName[]>;
}

export function useOptionsDetails(): UseOptionsDetailsHook {
    const allOptions = useSignal<InputOptionName[]>(createOptionNames());
    return { allOptions };
}