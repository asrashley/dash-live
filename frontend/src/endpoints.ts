import { createContext } from 'preact';
import log from 'loglevel';

import { routeMap, uiRouteMap} from '@dashlive/routemap';
import { CsrfTokenStore } from './CsrfTokenStore';
import { CsrfTokenCollection } from './types/CsrfTokenCollection';
import { AllManifests } from './types/AllManifests';
import { AllStreamsJson, AllStreamsResponse } from './types/AllStreams';
import { JWToken } from './types/JWToken';
import { MultiPeriodStream, MultiPeriodStreamJson } from './types/MultiPeriodStream';
import { DecoratedMultiPeriodStream } from "./types/DecoratedMultiPeriodStream";
import { MultiPeriodStreamSummary } from './types/MultiPeriodStreamSummary';
import { ContentRolesMap } from './types/ContentRolesMap';
import { ModifyMultiPeriodStreamJson, ModifyMultiPeriodStreamResponse } from './types/ModifyMultiPeriodStreamResponse';
import { InitialApiTokens } from './types/InitialApiTokens';
import { LoginRequest } from './types/LoginRequest';
import { LoginResponse } from './types/LoginResponse';
import { MultiPeriodStreamValidationRequest, MultiPeriodStreamValidationResponse } from './types/MpsValidation';
import { CgiOptionDescription } from './types/CgiOptionDescription';

type TokenStoreCollection = {
  files: CsrfTokenStore;
  kids: CsrfTokenStore;
  login: CsrfTokenStore;
  streams: CsrfTokenStore;
  upload: CsrfTokenStore;
};

export type ApiRequestOptions = {
  authorization?: string;
  body: RequestInit['body'];
  service: keyof TokenStoreCollection;
  signal?: AbortSignal;
  method: RequestInit['method'];
  rejectOnError?: boolean;
  query?: URLSearchParams;
};

export type GetAllStreamsProps = ApiRequestOptions & {
  withDetails: boolean
};

type RefreshAccessTokenResponse = {
  accessToken?: JWToken;
  csrfTokens: CsrfTokenCollection;
  ok: boolean;
  status: number;
};

export interface ApiRequestsProps extends Readonly<InitialApiTokens> {
  navigate: (url: string) => void;
}

export class ApiRequests {
  private csrfTokens: TokenStoreCollection;
  private accessToken: JWToken | null;
  private refreshToken: JWToken | null;
  private navigate: ApiRequestsProps['navigate'];

  constructor({csrfTokens, accessToken, refreshToken, navigate}: ApiRequestsProps) {
    this.csrfTokens = {
      files: new CsrfTokenStore(csrfTokens?.files),
      kids: new CsrfTokenStore(csrfTokens?.kids),
      login: new CsrfTokenStore(csrfTokens?.login),
      streams: new CsrfTokenStore(csrfTokens?.streams),
      upload: new CsrfTokenStore(csrfTokens?.upload),
    }
    this.accessToken = accessToken;
    this.refreshToken = refreshToken;
    this.navigate = navigate;
  }

  getAllManifests(options: Partial<ApiRequestOptions> = {}): Promise<AllManifests> {
    return this.sendApiRequest<AllManifests>(routeMap.listManifests.url(), options);
  }

  getContentRoles(options: Partial<ApiRequestOptions> = {}): Promise<ContentRolesMap> {
    return this.sendApiRequest(routeMap.contentRoles.url(), options);
  }

  getCgiOptions(options: Partial<ApiRequestOptions> = {}) : Promise<CgiOptionDescription[]> {
    return this.sendApiRequest(routeMap.cgiOptions.url(), options);
  }

  async getUserInfo(signal: AbortSignal): Promise<LoginResponse | Response> {
    if (!this.refreshToken) {
      throw new Error('Cannot get user info without a refresh token');
    }
    const options: Partial<ApiRequestOptions> = {
      authorization: this.refreshToken.jwt,
      method: 'GET',
      rejectOnError: false,
      signal,
    };
    const response = await this.sendApiRequest<LoginResponse | Response>(routeMap.login.url(), options);
    if (response['success']) {
      this.accessToken = (response as LoginResponse).accessToken;
    }
    return response;
  }

  async loginUser(request: LoginRequest, options: Partial<GetAllStreamsProps> = {}): Promise<LoginResponse> {
    const response: LoginResponse = await this.sendApiRequest<LoginResponse>(routeMap.login.url(), {
      body: JSON.stringify(request),
      method: 'POST',
      service: 'login',
      ...options,
    });
    if (response.success) {
      if (response.accessToken) {
        this.accessToken = response.accessToken;
      }
      if (response.refreshToken) {
        this.refreshToken = response.refreshToken;
      }
    }
    return response;
  }

  async logoutUser(options: Partial<GetAllStreamsProps> = {}): Promise<Response> {
    const response: Response = await this.sendApiRequest(routeMap.login.url(), {
      method: 'DELETE',
      service: 'login',
      ...options,
    });
    this.accessToken = null;
    this.refreshToken = null;
    return response;
  }

  async getAllStreams(options: Partial<GetAllStreamsProps> = {}): Promise<AllStreamsResponse> {
    const { streams, keys } = await this.sendApiRequest<AllStreamsJson>(routeMap.listStreams.url(), options);
    return { streams, keys};
  }

  getAllMultiPeriodStreams(options: Partial<ApiRequestOptions> = {}): Promise<MultiPeriodStreamSummary[]> {
    return this.sendApiRequest<MultiPeriodStreamSummary[]>(routeMap.listMps.url(), options);
  }

  async getMultiPeriodStream(mps_name: string, options: Partial<ApiRequestOptions> = {}): Promise<MultiPeriodStream> {
    const { model } = await this.sendApiRequest<MultiPeriodStreamJson>(routeMap.editMps.url({mps_name}), options);
    return model;
  }

  async addMultiPeriodStream(
      data: DecoratedMultiPeriodStream,
      options: Partial<ApiRequestOptions> = {}): Promise<ModifyMultiPeriodStreamResponse> {
    const service = 'streams';
    const csrf_token = await this.csrfTokens[service].getToken(
      options?.signal, this.getCsrfTokens);
    const { errors, success, model } = await this.sendApiRequest<ModifyMultiPeriodStreamJson>(routeMap.addMps.url(), {
      ...options,
      service,
      body: JSON.stringify({
        ...data,
        csrf_token
      }),
      method: 'PUT',
    });
    return { errors, success, model };
  }

  async modifyMultiPeriodStream(mps_name: string, data: DecoratedMultiPeriodStream,
     options: Partial<ApiRequestOptions> = {}): Promise<ModifyMultiPeriodStreamJson> {
    const service = 'streams';
    const csrf_token = await this.csrfTokens[service].getToken(
      options?.signal, this.getCsrfTokens);
    return await this.sendApiRequest<ModifyMultiPeriodStreamJson>(routeMap.editMps.url({mps_name}), {
      ...options,
      service,
      body: JSON.stringify({...data, csrf_token}),
      method: 'POST',
    });
  }

  async deleteMultiPeriodStream(mps_name: string, options: Partial<ApiRequestOptions> = {}): Promise<Response> {
    const service = 'streams';
    const csrf_token = await this.csrfTokens[service].getToken(
      options?.signal, this.getCsrfTokens
    );
    const query = new URLSearchParams({
      csrf_token,
    });
    return this.sendApiRequest<Response>(routeMap.editMps.url({mps_name}), {
      ...options,
      service,
      query,
      method: 'DELETE',
    });
  }

  async validateMultiPeriodStream(data: MultiPeriodStreamValidationRequest,
        options: Partial<ApiRequestOptions> = {}): Promise<MultiPeriodStreamValidationResponse> {
    const service = 'streams';
    const csrf_token = await this.csrfTokens[service].getToken(
      options?.signal, this.getCsrfTokens);
    return this.sendApiRequest<MultiPeriodStreamValidationResponse>(routeMap.validateMps.url(), {
      ...options,
      method: 'POST',
      service: 'streams',
      body: JSON.stringify({...data, csrf_token}),
    });
  }

  private async sendApiRequest<T>(url: string, options: Partial<ApiRequestOptions>): Promise<T> {
    const { authorization, body, service, signal, method='GET',
      rejectOnError = true } = options;
    let { query } = options;
    let usedAccessToken = false;
    const headers: Headers = new Headers({
      "Content-Type": "application/json",
    });
    if (authorization !== undefined) {
      headers.set('Authorization', `Bearer ${authorization}`);
    } else {
      if (!this.accessToken && this.refreshToken) {
        await this.getAccessToken(signal);
      }
      if (this.accessToken) {
        headers.set('Authorization', `Bearer ${this.accessToken.jwt}`);
        usedAccessToken = true;
      } else if (service) {
        const token = await this.csrfTokens[service].getToken(signal);
        if (query === undefined) {
          query = new URLSearchParams({csrf_token: token});
        } else {
          query.set('csrf_token', token);
        }
      }
    }
    if (query) {
      url = `${url}?${query.toString()}`;
    }
    if (signal && signal.aborted) {
      throw signal.reason;
    }
    const cache = 'no-cache';
    const credentials = 'same-origin';
    const mode = 'same-origin';
    url = new URL(url, document.location.origin).href;
    log.trace(`sendApiRequest(${url})`);
    let fetchResult = await fetch(url, {
      cache,
      credentials,
      mode,
      body,
      headers,
      method,
      signal,
    });
    log.trace(`url=${url} status=${fetchResult.status} usedAccessToken=${usedAccessToken}`);
    if (fetchResult.status === 401 && usedAccessToken && this.refreshToken) {
      await this.getAccessToken(signal);
      headers.set('Authorization', `Bearer ${this.accessToken.jwt}`);
      log.trace(`retrying sendApiRequest(${url})`);
      fetchResult = await fetch(url, {
          cache,
          credentials,
          mode,
          body,
          headers,
          method,
          signal,
      });
    }
    if (!fetchResult.ok) {
      if (rejectOnError) {
        throw new Error(`${ url }: ${ fetchResult.status }`);
      } else {
        return fetchResult as T;
      }
    }
    if (fetchResult.status !== 200) {
      return fetchResult as T;
    }
    const data = await fetchResult.json();
    if (signal?.aborted) {
      throw signal.reason;
    }
    if (typeof(data?.csrfTokens) === "object") {
      this.updateCsrfTokens(data.csrfTokens);
    } else if(typeof(data?.csrf_tokens) === "object") {
      this.updateCsrfTokens(data.csrf_tokens);
    } else if (typeof(data?.csrf_token) === "string") {
      this.updateCsrfTokens({[service]: data.csrf_token});
    }

    return data as T;
  }

  private updateCsrfTokens(tokens: Partial<CsrfTokenCollection>) {
    for (const [key, value] of Object.entries(tokens)) {
      if (this.csrfTokens[key] === undefined) {
        this.csrfTokens[key] = new CsrfTokenStore(value);
      } else {
        this.csrfTokens[key].setToken(value);
      }
    }
  }

  private async getAccessToken(signal: AbortSignal): Promise<JWToken> {
    if (!this.refreshToken) {
      throw new Error('Cannot request an access token without a refresh token');
    }
    const options = {
      authorization: this.refreshToken.jwt,
      rejectOnError: false,
      signal,
    };
    const data = await this.sendApiRequest<RefreshAccessTokenResponse>(routeMap.refreshAccessToken.url(), options);
    const { accessToken, csrfTokens, ok, status } = data ?? {};
    if (ok === false && status === 401) {
      this.navigate(uiRouteMap.login.url());
    }
    if (!accessToken) {
      throw new Error('Failed to refresh access token');
    }
    this.accessToken = accessToken;
    if (csrfTokens) {
      this.updateCsrfTokens(csrfTokens);
    }
    return accessToken;
  }

  private getCsrfTokens = async (signal: AbortSignal): Promise<void> => {
    if (!this.accessToken && this.refreshToken) {
      await this.getAccessToken(signal);
    }
    if (!this.accessToken) {
      throw new Error('Cannot request CSRF tokens without an access token');
    }
    const options = {
      authorization: this.accessToken.jwt,
      signal,
    };
    const data = await this.sendApiRequest<RefreshAccessTokenResponse>(routeMap.refreshCsrfTokens.url(), options);
    const { csrfTokens } = data ?? {};
    if (csrfTokens) {
      this.updateCsrfTokens(csrfTokens);
    }
  };
}

export const EndpointContext = createContext<ApiRequests>(null);
