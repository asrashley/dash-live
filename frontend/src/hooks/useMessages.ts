import { type ReadonlySignal, signal, useComputed } from "@preact/signals";
import { MessageLevel, MessageType } from "../types/MessageType";

type MessagesState = {
  alerts: MessageType[];
  nextId: number;
}

const messages = signal<MessagesState>({ alerts: [], nextId: 1 });

// only exported for use in tests
export function resetAllMessages() {
  messages.value = { alerts: [], nextId: 1 };
}

export type AppendMessageFn = (level: MessageLevel, text: string) => number;

const appendMessage: AppendMessageFn = (level: MessageLevel, text: string) => {
  const id = messages.value.nextId;
  const alerts = [
    ...messages.value.alerts,
    {
      text,
      id,
      level,
    },
  ];
  messages.value = {
    nextId: id + 1,
    alerts,
  };
  return id;
};

const removeAlert = (id: number) => {
  const alerts = messages.value.alerts.filter((a) => a.id !== id);
  messages.value = { ...messages.value, alerts };
};

export interface UseMessagesHook {
  alerts: ReadonlySignal<MessageType[]>
  appendMessage: AppendMessageFn;
  removeAlert: (id: number) => void;
}

export function useMessages(): UseMessagesHook {
  const alerts = useComputed(() => {
    return messages.value.alerts;
  });

  return { alerts, appendMessage, removeAlert };
}
