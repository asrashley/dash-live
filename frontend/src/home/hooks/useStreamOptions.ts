import { useCallback } from "preact/hooks";
import { type ReadonlySignal, type Signal, useComputed, useSignal } from "@preact/signals";

import { defaultCgiOptions, drmSystems } from "@dashlive/options";
import { CombinedStream } from "../../hooks/useCombinedStreams";
import { InputFormData } from "../../types/InputFormData";

const keyName = "dashlive.homepage.options";

function getDefaultOptions(): InputFormData {
  const lsKey = localStorage.getItem(keyName);
  const previousOptions = lsKey ? JSON.parse(lsKey) : {};
  return {
    ...defaultCgiOptions,
    manifest: "hand_made.mpd",
    mode: "vod",
    stream: undefined,
    ...previousOptions,
  };
}

const skipKeys = new RegExp(`^(${[
  ...drmSystems,
  ...drmSystems.map(name => `${name}__enabled`),
  ...drmSystems.map(name => `${name}__drmloc`),
  'drms', 'manifest', 'stream', 'mode'].join('|')})$`);

const manifestSkipKeys = /^(player|dashjs|shaka)/;

export type EnabledDrmSystems = {
  [name: string]: boolean;
};

export interface UseStreamOptionsHook {
  data: Signal<InputFormData>;
  drms: ReadonlySignal<EnabledDrmSystems>;
  stream: ReadonlySignal<CombinedStream>;
  mode: ReadonlySignal<string>;
  manifest: ReadonlySignal<string>;
  nonDefaultOptions: ReadonlySignal<object>;
  manifestOptions: ReadonlySignal<object>;
  setValue: (name: string, value: string | number | boolean) => void;
  resetAllValues: () => void;
}

const emptyStream: CombinedStream = {
  title: "",
  value: "",
  mps: false
}

export interface UseStreamOptionsProps {
  streamNames: ReadonlySignal<string[]>;
  streamsMap: ReadonlySignal<Map<string, CombinedStream>>;
}
export function useStreamOptions({ streamNames, streamsMap }: UseStreamOptionsProps): UseStreamOptionsHook {
  const data = useSignal<InputFormData>(getDefaultOptions());
  const stream = useComputed<CombinedStream>(() => {
    const name: string = (data.value['stream'] as string | undefined) ?? streamNames.value[0];
    return streamsMap.value.get(name) ?? emptyStream;
  });
  const mode = useComputed<string>(() => data.value['mode'] as string);
  const manifest = useComputed<string>(() => data.value['manifest'] as string);
  const drms = useComputed<EnabledDrmSystems>(() => Object.fromEntries(drmSystems.map(name => [name, data.value[name] === "1"])));
  const nonDefaultOptions = useComputed<object>(() => {
    const params = Object.entries(data.value)
      .filter(([key, value]) => defaultCgiOptions[key] != value)
      .filter(([key]) => !skipKeys.test(key));
    const drm = drmSystems.filter(system => data.value[system] === "1").map(system => {
      const drmLoc = data.value[`${system}__drmloc`];
      return drmLoc ? `${system}-${drmLoc}`: system;
    });
    if (drm.length) {
      params.push(['drm', drm.join(',')]);
    }
    return Object.fromEntries(params);
  });
  const manifestOptions = useComputed<object>(() => Object.fromEntries(
    Object.entries(nonDefaultOptions.value).filter(([key]) => !manifestSkipKeys.test(key))));

  const setValue = useCallback(
    (name: string, value: string | number | boolean) => {
      if (value === true) {
        value = "1";
      } else if (value === false) {
        value = "0";
      }
      data.value = {
        ...data.value,
        [name]: value,
      };
      if (name === "stream" && mode.value === "odvod") {
        const nextStream = streamsMap.value.get(value as string);
        if (nextStream?.mps) {
          data.value = {
            ...data.value,
            mode: "vod",
          };
        }
      }
      const params = Object.fromEntries(
        Object.entries(data.value).filter(
          ([key, value]) => defaultCgiOptions[key] !== value
        )
      );
      localStorage.setItem(keyName, JSON.stringify(params));
    },
    [data, mode, streamsMap]
  );

  const resetAllValues = useCallback(() => {
    data.value = { ...getDefaultOptions() };
    localStorage.removeItem(keyName);
  }, [data]);

  return { data, drms, stream, mode, manifest, nonDefaultOptions, manifestOptions, setValue, resetAllValues };
}
