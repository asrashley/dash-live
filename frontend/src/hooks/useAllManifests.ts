import { createContext } from "preact";
import { useContext, useEffect } from "preact/hooks";
import { Signal, useComputed, useSignal } from "@preact/signals";

import { EndpointContext } from "../endpoints";
import { AllManifests } from "../types/AllManifests";


export interface UseAllManifestsHooks {
  allManifests: Signal<AllManifests | undefined>;
  names: Signal<string[]>;
  error: Signal<string | null>;
}

export const AllManifestsContext = createContext<UseAllManifestsHooks>(null);

export function useAllManifests(): UseAllManifestsHooks {
  const apiRequests = useContext(EndpointContext);
  const allManifests = useSignal<AllManifests | undefined>();
  const error = useSignal<string | null>(null);
  const names = useComputed<string[]>(() => {
    const keys = allManifests.value ? [...Object.keys(allManifests.value)]: [];
    keys.sort();
    return keys;
  });

  useEffect(() => {
    const controller = new AbortController();
    const { signal } = controller;

    const fetchAllManifestsIfRequired = async () => {
      if (!allManifests.value) {
        try {
          const data = await apiRequests.getAllManifests({
            signal,
          });
          if (!signal.aborted) {
            allManifests.value = data;
            error.value = null;
          }
        } catch (err) {
          if (!signal.aborted) {
            console.error(err);
            error.value = `${err}`;
          }
        }
      }
    };
    fetchAllManifestsIfRequired();

    return () => {
      controller.abort();
    };
  }, [apiRequests, allManifests, error]);

  return { allManifests, names, error };
}
