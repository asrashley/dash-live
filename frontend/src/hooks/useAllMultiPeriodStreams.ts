import { createContext } from "preact";
import { useEffect, useState, useContext, useCallback } from "preact/hooks";
import { type ReadonlySignal, useSignal } from "@preact/signals";

import { EndpointContext } from "../endpoints";
import { AllMultiPeriodStreamsJson, MultiPeriodStreamSummary } from "../types/AllMultiPeriodStreams";

export interface UseAllMultiPeriodStreamsHook {
  error: ReadonlySignal<string | null>;
  streams: ReadonlySignal<MultiPeriodStreamSummary[]>;
  loaded: ReadonlySignal<boolean>;
  sort: (field: string, ascending: boolean) => void;
  sortField: string;
  sortAscending: boolean;
}

export const AllMultiPeriodStreamsContext = createContext<UseAllMultiPeriodStreamsHook>(null);

function sortStreams(streams: ReadonlySignal<MultiPeriodStreamSummary[]>, field: string, ascending: boolean): MultiPeriodStreamSummary[] {
  const newOrder = [...streams.value];
  newOrder.sort((a, b) => {
    const left = a[field];
    const right = b[field];
    if (left === right) {
      return 0;
    }
    if (left < right) {
      return ascending ? -1 : 1;
    }
    return ascending ? 1 : -1;
  });
  return newOrder;
}

export function useAllMultiPeriodStreams(): UseAllMultiPeriodStreamsHook {
  const apiRequests = useContext(EndpointContext);
  const streams = useSignal<MultiPeriodStreamSummary[]>([]);
  const loaded = useSignal<boolean>(false);
  const error = useSignal<string | null>(null);
  const [sortField, setSortField] = useState<string>("name");
  const [sortAscending, setSortAscending] = useState<boolean>(true);

  const sort = useCallback(
    (field: string, ascending: boolean) => {
      setSortField(field);
      setSortAscending(ascending);
      streams.value = sortStreams(streams, field, ascending);
    },
    [streams]
  );

  useEffect(() => {
    const controller = new AbortController();
    const { signal } = controller;

    const fetchStreamsIfRequired = async () => {
      if (!loaded.value) {
        try {
          const data: AllMultiPeriodStreamsJson = await apiRequests.getAllMultiPeriodStreams({ signal });
          if (!signal.aborted) {
            loaded.value = true;
            streams.value = data.streams;
            error.value = null;
            sortStreams(streams, sortField, sortAscending);
          }
        } catch (err) {
          if (!signal.aborted) {
            error.value = `Fetching multi-period streams list failed: ${err}`;
          }
        }
      }
    };

    fetchStreamsIfRequired();

    return () => {
      controller.abort();
    };
  }, [apiRequests, error, loaded, sortAscending, sortField, streams]);

  return { error, streams, loaded, sort, sortField, sortAscending };
}
