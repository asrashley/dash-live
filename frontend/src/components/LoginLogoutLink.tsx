import { useCallback, useContext, useState } from "preact/hooks";
import { useLocation } from "wouter-preact";

import { uiRouteMap } from "@dashlive/routemap";

import { EndpointContext } from "../endpoints";
import { useMessages } from "../hooks/useMessages";
import { WhoAmIContext } from "../hooks/useWhoAmI";
import { Icon } from "./Icon";

export function LoginLogoutLink() {
  const { user, setUser } = useContext(WhoAmIContext);
  const api = useContext(EndpointContext);
  const { appendMessage } = useMessages();
  const setLocation = useLocation()[1];
  const [expanded, setExpanded] = useState<boolean>(false);
  const onLogin = useCallback(() => {
    setLocation(uiRouteMap.login.url());
  }, [setLocation]);
  const onChangePassword = useCallback(() => {
    setExpanded(false);
    setLocation(uiRouteMap.changePassword.url());
  }, [setLocation]);
  const onLogOut = useCallback(async () => {
    setExpanded(false);
    setUser(null);
    try {
      await api.logoutUser();
      appendMessage("success", "You have successfully logged out");
    } catch (err) {
      appendMessage("danger", `Logout failed: ${err}`);
    }
    setLocation(uiRouteMap.home.url());
  }, [setUser, setLocation, api, appendMessage]);
  const toggleExpand = useCallback(() => {
    setExpanded(!expanded);
  }, [expanded]);
  const toggleClasses = `btn btn-light dropdown-toggle${
    expanded ? " show" : ""
  }`;
  const listClasses = `dropdown-menu${expanded ? " show" : ""}`;

  if (!user.value.isAuthenticated) {
    return (
      <li className="nav-item dropdown ms-auto me-4">
        <button className="nav-link" role="menuitem" onClick={onLogin}>
          Log In
        </button>
      </li>
    );
  }

  return (
    <li className="nav-item dropdown ms-auto me-4">
      <span className="username me-1">{user.value.username}</span>
      <button
        className={toggleClasses}
        role="menubar"
        data-testid="toggle-user-menu"
        onClick={toggleExpand}
        aria-expanded={expanded}>
        <Icon name="gear-fill" />
      </button>
      <ul className={listClasses}>
        <li>
          <button className="dropdown-item" role="menuitem" onClick={onChangePassword}>
            Change Password
          </button>
        </li>
        <li>
          <button className="dropdown-item" role="menuitem" onClick={onLogOut}>
            Log Out
          </button>
        </li>
      </ul>
    </li>
  );
}
