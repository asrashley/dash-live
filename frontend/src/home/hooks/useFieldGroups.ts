import { useContext } from "preact/hooks";
import { type ReadonlySignal, useComputed } from "@preact/signals";

import { fieldGroups, drmSystems } from '@dashlive/options';

import { useAllManifests  } from '../../hooks/useAllManifests';
import { UseCombinedStreams  } from '../../hooks/useCombinedStreams';
import { StaticInputProps } from "../../form/types/StaticInputProps";
import { InputFormGroup } from "../../form/types/InputFormGroup";

const drmSkipNames = new RegExp(`^${drmSystems.map(name => `${name}__enabled`).join('|')}$`);

interface UseFieldGroupsHook {
  homeFieldGroups: ReadonlySignal<InputFormGroup[]>;
}

export function useFieldGroups(): UseFieldGroupsHook {
  const { allManifests, names } = useAllManifests();
  const { streamNames, streamsMap } = useContext(UseCombinedStreams);

  const homeFieldGroups = useComputed<InputFormGroup[]>(() => {
    const playbackMode: StaticInputProps = {
      name: "mode",
      shortName: "mode",
      fullName: "playbackMode",
      title: "Playback Mode",
      type: "radio",
      options: [
        {
          title: "Video On Demand (using live profile)",
          value: "vod",
        },
        {
          title: "Live stream (using live profile)",
          value: "live",
        },
        {
          title: "Video On Demand (using on-demand profile)",
          value: "odvod",
        },
      ],
    };
    const selectManifest: StaticInputProps = {
      name: "manifest",
      shortName: "manifest",
      fullName: "manifest",
      title: "Manifest",
      text: "Manifest template to use",
      type: "select",
      options: names.value.map((name: string) => {
        const msft = allManifests.value[name];
        return {
          title: msft.title,
          value: name,
        };
      }),
    };
    const selectStream: StaticInputProps = {
      name: "stream",
      shortName: "stream",
      fullName: "stream",
      title: "Stream",
      text: "Stream to play",
      type: "select",
      options: streamNames.value.map((value: string) => ({
        title: streamsMap.value.get(value).title,
        value,
      })),
    };
    const selectDrmSystem: StaticInputProps = {
      name: "drms",
      shortName: "drms",
      fullName: "drms",
      title: "DRM systems",
      text: "DRM systems to enable for encrypted AdaptationSets",
      type: "multiselect",
      options: drmSystems.map((name: string) => ({
        name,
        title: name,
        value: name,
      })),
    };
    const generalOptions: InputFormGroup = {
      ...fieldGroups[0],
      fields: [
        playbackMode,
        selectManifest,
        selectStream,
        selectDrmSystem,
        ...fieldGroups[0].fields,
      ],
    };
    const otherGroups: InputFormGroup[] = fieldGroups.slice(1).map(grp => ({
        ...grp,
        fields: grp.fields.filter(({name}) => !drmSkipNames.test(name)),
    }));
    return [generalOptions, ...otherGroups];
  });

  return { homeFieldGroups };
}
