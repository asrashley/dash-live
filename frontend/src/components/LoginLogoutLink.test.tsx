import { afterEach, beforeEach, describe, expect, test, vi } from "vitest";
import { mock } from "vitest-mock-extended";
import { useLocation } from "wouter-preact";

import { uiRouteMap } from "@dashlive/routemap";
import { renderWithProviders } from "../test/renderWithProviders";
import { ApiRequests, EndpointContext } from "../endpoints";
import { useMessages, UseMessagesHook } from "../hooks/useMessages";
import { normalUser } from "../test/MockServer";

import { LoginLogoutLink } from "./LoginLogoutLink";
import userEvent from "@testing-library/user-event";

vi.mock("wouter-preact", async (importOriginal) => {
  return {
    ...(await importOriginal()),
    useLocation: vi.fn(),
  };
});

vi.mock("../hooks/useMessages", async (importOriginal) => {
  return {
    ...(await importOriginal()),
    useMessages: vi.fn(),
  };
});

describe("LoginLogoutLink component", () => {
  const useLocationMock = vi.mocked(useLocation);
  const useMessagesMock = vi.mocked(useMessages);
  const setLocation = vi.fn();
  const apiRequests = mock<ApiRequests>();
  const messagesMock = mock<UseMessagesHook>();

  beforeEach(() => {
    useLocationMock.mockReturnValue(["/", setLocation]);
    useMessagesMock.mockReturnValue(messagesMock);
  });

  afterEach(() => {
    vi.clearAllMocks();
  });

  test("shows login link", async () => {
    const user = userEvent.setup();
    const { asFragment, getByText } = renderWithProviders(
      <EndpointContext.Provider value={apiRequests}>
        <LoginLogoutLink />
      </EndpointContext.Provider>
    );
    const btn = getByText("Log In") as HTMLButtonElement;
    await user.click(btn);
    expect(setLocation).toHaveBeenCalledTimes(1);
    expect(setLocation).toHaveBeenCalledWith(uiRouteMap.login.url());
    expect(asFragment()).toMatchSnapshot();
  });

  test("shows log out link", async () => {
    const user = userEvent.setup();
    const prom = new Promise<void>((resolve) => {
      apiRequests.logoutUser.mockImplementationOnce(() => {
        resolve();
        return Promise.resolve(
          new Response(null, {
            status: 204,
          })
        );
      });
    });

    const { asFragment, getByTestId, findByText, whoAmI } =
      renderWithProviders(
        <EndpointContext.Provider value={apiRequests}>
          <LoginLogoutLink />
        </EndpointContext.Provider>,
        { userInfo: normalUser }
    );
    const toggle = getByTestId("toggle-user-menu") as HTMLButtonElement;
    await user.click(toggle);
    const btn = await findByText("Log Out") as HTMLButtonElement;
    expect(whoAmI.user.value.isAuthenticated).toEqual(true);
    await user.click(btn);
    await prom;
    expect(whoAmI.user.value.isAuthenticated).toEqual(false);
    expect(setLocation).toHaveBeenCalledTimes(1);
    expect(setLocation).toHaveBeenCalledWith(uiRouteMap.home.url());
    await findByText("Log In");
    expect(messagesMock.appendMessage).toHaveBeenCalledTimes(1);
    expect(messagesMock.appendMessage).toHaveBeenCalledWith(
      "success",
      "You have successfully logged out"
    );
    expect(asFragment()).toMatchSnapshot();
  });

  test("shows error if log out fails", async () => {
    const user = userEvent.setup();
    const prom = new Promise<void>((resolve) => {
      apiRequests.logoutUser.mockImplementationOnce(() => {
        resolve();
        return Promise.reject(new Error("500: Server Error"));
      });
    });

    const { findByText, getByTestId, whoAmI } = renderWithProviders(
      <EndpointContext.Provider value={apiRequests}>
        <LoginLogoutLink />
      </EndpointContext.Provider>,
      { userInfo: normalUser }
    );
    const toggle = getByTestId("toggle-user-menu") as HTMLButtonElement;
    await user.click(toggle);
    const btn = await findByText('Log Out') as HTMLButtonElement;
    await user.click(btn);
    await prom;
    expect(whoAmI.user.value.isAuthenticated).toEqual(false);
    expect(setLocation).toHaveBeenCalledTimes(1);
    expect(setLocation).toHaveBeenCalledWith(uiRouteMap.home.url());
    await findByText("Log In");
    expect(messagesMock.appendMessage).toHaveBeenCalledTimes(1);
    expect(messagesMock.appendMessage).toHaveBeenCalledWith(
      "danger",
      "Logout failed: Error: 500: Server Error"
    );
  });
});
