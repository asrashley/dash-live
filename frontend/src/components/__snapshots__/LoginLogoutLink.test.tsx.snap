// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`LoginLogoutLink component > shows  log out link 1`] = `
<DocumentFragment>
  <li
    class="nav-item user-login"
  >
    <a
      class="nav-link user-login"
      href="/login"
    >
      Log In
    </a>
  </li>
</DocumentFragment>
`;

exports[`LoginLogoutLink component > shows login link 1`] = `
<DocumentFragment>
  <li
    class="nav-item user-login"
  >
    <a
      class="nav-link user-login"
      href="/login"
    >
      Log In
    </a>
  </li>
</DocumentFragment>
`;
